File: src/tools/important/tankslua/TanksLib.kt
package tools.important.tankslua

import party.iroiro.luajava.JFunction
import party.iroiro.luajava.Lua
import tanks.Drawing
import tanks.Game
import tanks.ModAPI
import tools.important.tankslua.gui.Notification
import tools.important.tankslua.gui.NotificationType

fun openTanksLib(luaState: Lua) {
    luaState.createTable(0, 0)
    val tanksLib = luaState.get()

    tanksLib.set("notify", JFunction(fun(state): Int {
        val argCount = state.top

        if (argCount < 1) luaError("Not enough arguments!")

        val message: String = state.toString(1) ?: luaError("Message is wrong type!")

        val duration: Double? = if (argCount >= 2) {
            if (!state.isNumber(2)) luaError("Duration is wrong type!")
            state.toNumber(2)
        } else null

        Notification(message, NotificationType.INFO, duration ?: 200.0)
        return 0
    }))

    tanksLib.set("version", Game.version)

    tanksLib.set("isModApi", try {
        ModAPI::class.java.getDeclaredField("version")
        true
    } catch (_: NoSuchFieldException) {
        false
    })

    luaState.createTable(0, 0)
    val tanksDrawingLib = luaState.get()

    tanksDrawingLib.set("color", JFunction(fun(state): Int {
        val argCount = state.top
        if (argCount < 3) luaError("Not enough arguments!")

        val r = if (state.isNumber(1)) state.toNumber(1) else luaError("R is wrong type!")
        val g = if (state.isNumber(2)) state.toNumber(2) else luaError("G is wrong type!")
        val b = if (state.isNumber(3)) state.toNumber(3) else luaError("B is wrong type!")

        val opacity: Double? = if (argCount >= 4) {
            if (state.isNumber(4)) state.toNumber(4) else luaError("Opacity is wrong type!")
        } else null

        state.pop(3 +
                (if (opacity != null) 1 else 0)
        )

        Drawing.drawing.setColor(r,g,b,opacity ?: 255.0)

        return 0
    }))

    tanksDrawingLib.set("size", JFunction(fun(state): Int {
        val argCount = state.top

        val interfaceMode = if (argCount >= 1) {
            if (state.isBoolean(1)) state.toBoolean(1) else luaError("Interface mode is wrong type!")
        } else null
        state.pop(1)

        if (interfaceMode == true) {
            state.push(Drawing.drawing.interfaceSizeX)
            state.push(Drawing.drawing.interfaceSizeY)
            return 2
        }
        state.push(Drawing.drawing.sizeX)
        state.push(Drawing.drawing.sizeY)
        return 2
    }))

    tanksDrawingLib.set("oval", JFunction(fun(state): Int {
        val argCount = state.top
        if (argCount < 6) luaError("Not enough arguments!")

        val interfaceMode = if (state.isBoolean(1)) state.toBoolean(2) else luaError("Interface mode is wrong type!")
        val x = if (state.isNumber(2)) state.toNumber(2) else luaError("X is wrong type!")
        val y = if (state.isNumber(3)) state.toNumber(3) else luaError("Y is wrong type!")
        val width = if (state.isNumber(4)) state.toNumber(4) else luaError("Width is wrong type!")
        val height = if (state.isNumber(5)) state.toNumber(5) else luaError("Height is wrong type!")
        val drawMode = if (state.isString(6)) state.toString(6) else luaError("Draw mode is wrong type!")

        state.pop(6)

        if (interfaceMode) {
            if (drawMode == "fill") {
                Drawing.drawing.fillInterfaceOval(x,y,width,height)
            } else if (drawMode == "line") {
                Drawing.drawing.drawInterfaceOval(x,y,width,height)
            }
        } else {
            if (drawMode == "fill") {
                Drawing.drawing.fillOval(x,y,width,height)
            } else if (drawMode == "line") {
                Drawing.drawing.drawOval(x,y,width,height)
            }
        }

        return 0
    }))

    tanksDrawingLib.set("fontSize", JFunction(fun(state): Int {
        val argCount = state.top

        val forInterface = if (argCount >= 1) {
            if (state.isBoolean(1)) state.toBoolean(1) else luaError("Interface mode is wrong type!")
        } else null

        val newSize = if (argCount >= 2) {
            if (state.isNumber(2)) state.toNumber(2) else luaError("New size is wrong type!")
        } else null

        state.pop(
            (if (forInterface != null) 1 else 0) +
                    (if (newSize != null) 1 else 0)
        )

        if (newSize == null) {
            if (forInterface == true) {
                state.push(Drawing.drawing.fontSize / 36.0 * Drawing.drawing.interfaceScale)
            } else {
                state.push(Drawing.drawing.fontSize)
            }
            return 1
        }

        if (forInterface == true) {
            Drawing.drawing.setInterfaceFontSize(newSize)
        } else {
            Drawing.drawing.setFontSize(newSize)
        }

        return 0
    }))

    tanksDrawingLib.set("text", JFunction(fun(state): Int {
        val argCount = state.top

        if (argCount < 4) luaError("Not enough arguments!")

        val forInterface = if (state.isBoolean(1)) state.toBoolean(1) else luaError("Interface mode is wrong type!")
        val x = if (state.isNumber(2)) state.toNumber(2) else luaError("X is wrong type!")
        val y = if (state.isNumber(3)) state.toNumber(3) else luaError("Y is wrong type!")
        val text = if (state.isString(4)) state.toString(4) else luaError("Text is wrong type!")

        if (forInterface) {
            Drawing.drawing.drawInterfaceText(x, y, text)
        } else {
            Drawing.drawing.drawText(x, y, text)
        }

        return 0
    }))

    tanksDrawingLib.set("objYSpace", Drawing.drawing.objYSpace)
    tanksDrawingLib.set("objXSpace", Drawing.drawing.objXSpace)

    tanksLib.set("drawing", tanksDrawingLib)

    luaState.set("tanks", tanksLib)
}

File: resources/lua/tankslib.lua
tanks = {}

local Notification = java.import("tools.important.tankslua.gui.Notification")
local NotificationType = java.import("tools.important.tankslua.gui.NotificationType")
local Drawing = java.import("tanks.Drawing")
local Game = java.import("tanks.Game")

function tanks.notify(message, duration)
    Notification(message, NotificationType.INFO, duration or 200)
end

function tanks.playSound(sound, pitch)
    Drawing.drawing:playSound(sound, pitch or 1)
end



tanks.game = {}
tanks.game.version = Game.version
tanks.game.isModApi = pcall(function()
    java.import("tanks.ModAPI"):getDeclaredField("version")
end)


tanks.game.tileSize = Game.tile_size

function tanks.game.distance(a, b)
    return math.sqrt(tanks.game.sqDistance(a, b))
end

function tanks.game.sqDistance(a, b)
    return (a.posX - b.posX)^2 + (a.posY - b.posY)^2
end

function tanks.game.getMovables()
    return java.luaify(Game.movables)
end

function tanks.game.getPlayerTank()
    return Game.playerTank
end

function tanks.game.playing()
    return
        java.import("tanks.gui.screen.ScreenGame").class:isInstance(Game.screen)
        and not Game.screen.paused
        and Game.screen.playing
        and not Game.playerTank.destroy
end

function tanks.game.allied(a, b)
    return a.team ~= nil and b.team ~= nil and a.team == b.team
end



tanks.input = {}

tanks.input.codes = {}
for _, field in pairs(
    java.luaify(
        java.import("basewindow.InputCodes").class:getFields()
    )
) do
    tanks.input.codes[field:getName()] = field:get(nil)
end

function tanks.input.keyHeld(keyToCheck)
    local keys = java.luaify(Game.game.window.validPressedKeys)
    for _, key in pairs(keys) do
        if key == keyToCheck then return true end
    end
    return false
end

function tanks.input.mouseButtonHeld(buttonToCheck)
    local buttons = java.luaify(Game.game.window.validPressedButtons)
    for _, button in pairs(buttons) do
        if button == buttonToCheck then return true end
    end
    return false
end



tanks.drawing = {}

tanks.drawing.objWidth = Drawing.drawing.objWidth
tanks.drawing.objHeight = Drawing.drawing.objHeight
tanks.drawing.objXSpace = Drawing.drawing.objXSpace
tanks.drawing.objYSpace = Drawing.drawing.objYSpace

function tanks.drawing.color(r, g, b, a)
    if not a then
        Drawing.drawing:setColor(r, g, b)
        return
    end
    Drawing.drawing:setColor(r, g, b, a)
end


function tanks.drawing.oval(x, y, width, height, drawMode, coordinateMode)
    if coordinateMode == "interface" then
        if drawMode == "fill" then
            Drawing.drawing:fillInterfaceOval(x, y, width, height)
        elseif drawMode == "line" then
            Drawing.drawing:drawInterfaceOval(x, y, width, height)
        end
        return
    end

    if coordinateMode == "game" then
        if drawMode == "fill" then
            Drawing.drawing:fillOval(x, y, width, height)
        elseif drawMode == "line" then
            Drawing.drawing:drawOval(x, y, width, height)
        end
        return
    end
end

function tanks.drawing.rect(x, y, width, height, drawMode, coordinateMode)
    if coordinateMode == "interface" then
        if drawMode == "fill" then
            Drawing.drawing:fillInterfaceRect(x, y, width, height)
        elseif drawMode == "line" then
            Drawing.drawing:drawInterfaceRect(x, y, width, height)
        end
        return
    end

    if coordinateMode == "game" then
        if drawMode == "fill" then
            Drawing.drawing:fillRect(x, y, width, height)
        elseif drawMode == "line" then
            Drawing.drawing:drawRect(x, y, width, height)
        end
        return
    end
end

function tanks.drawing.fontSize(newSize, coordinateMode)
    if not newSize then -- get
        if coordinateMode == "interface" then
            return Drawing.drawing.fontSize / 36.0 * Drawing.drawing.interfaceScale
        end
        if coordinateMode == "game" then
            return Drawing.drawing.fontSize
        end
        return nil
    end

    -- set
    if coordinateMode == "interface" then
        Drawing.drawing:setInterfaceFontSize(newSize)
        return
    end
    if coordinateMode == "game" then
        Drawing.drawing:setFontSize(newSize)
        return
    end

    return nil
end

function tanks.drawing.text(x, y, text, coordinateMode)
    if coordinateMode == "interface" then
        Drawing.drawing:drawInterfaceText(x, y, text)
        return
    end

    if coordinateMode == "game" then
        Drawing.drawing:drawText(x, y, text)
        return
    end
end

function tanks.drawing.size(coordinateMode)
    if coordinateMode == "interface" then
        return Drawing.drawing.interfaceSizeX, Drawing.drawing.interfaceSizeY
    end

    if coordinateMode == "game" then
        return Drawing.drawing.sizeX, Drawing.drawing.sizeY
    end
end



tanks.multiplayer = {}

function tanks.multiplayer.isServer()
    return java.import("tanks.gui.screen.ScreenPartyHost").isServer
end

function tanks.multiplayer.isClient()
    return java.import("tanks.gui.screen.ScreenPartyLobby").isClient
end
